crates/solver/src/settlement_access_list.rs

```diff
@@ -1,37 +1,123 @@
+/// particular, our settlement contract will fail if the receiver is a smart contract. Because of
+/// this, if the receiver is a smart contract and we try to estimate the access list, the access
+/// list estimation will also fail.
+///
+/// The reason why this failure happens is because the Ethereum protocol sets a hard gas limit on
+/// transferring ETH into a smart contract, which some contracts exceed unless the access list is
+/// already specified.
+///
+/// The solution is to do access list estimation in two steps: first, simulate moving 1 wei
+/// into every smart contract to get a partial access list, and then simulate the full access list,
+/// passing the partial access list into the simulation. This way the settlement contract does not
+/// fail, and hence the full access list estimation also does not fail.
+pub async fn estimate_settlement_access_list(
+    estimating: &dyn AccessListEstimating,
+    web3: Web3,
+    solver_account: Account,
+    settlement: &Settlement,
+    tx: &TransactionBuilder<DynTransport>,
+) -> Result<AccessList> {
+    // Generate partial access lists for all smart contracts
+    let partial_access_lists = try_join_all(settlement.trades().map(|trade| async {
+        let buy_token = trade.order.data.buy_token;
+        let receiver = trade
+            .order
+            .data
+            .receiver
+            .unwrap_or(trade.order.metadata.owner);
+        let partial_access_list =
+            if buy_token == BUY_ETH_ADDRESS && is_smart_contract(&web3, receiver).await? {
+                estimating
+                    .estimate_access_list(
+                        &TransactionBuilder::new(web3.clone())
+                            .from(solver_account.clone())
+                            .to(receiver)
+                            .value(1.into()),
+                        None,
+                    )
+                    .await?
+            } else {
+                Default::default()
+            };
+        Result::<_>::Ok(partial_access_list)
+    }))
+    .await?;
+
+    // Merge the partial access lists together
+    let mut partial_access_list: HashMap<H160, HashSet<H256>> = Default::default();
+    for access_list in partial_access_lists {
+        for item in access_list {
+            partial_access_list
+                .entry(item.address)
+                .or_default()
+                .extend(item.storage_keys.into_iter());
+        }
+    }
+    let partial_access_list = partial_access_list
+        .into_iter()
+        .map(|(address, storage_keys)| AccessListItem {
+            address,
+            storage_keys: storage_keys.into_iter().collect(),
+        })
+        .collect_vec();
+
+    // Generate the final access list
+    estimating
+        .estimate_access_list(tx, Some(partial_access_list))
+        .await
+}
+
+async fn is_smart_contract(web3: &Web3, address: H160) -> Result<bool> {
+    Ok(!web3.eth().code(address, None).await?.0.is_empty())
+}
```

Comment:
We have a `CodeFetching` component that does something similar and might be worth reusing here (it does things like caching which would be useful here).